# Compile Twitter Data
Compile raw .json files from twitter feed into an organized file for performing analytics 

In [1]:
import json
import matplotlib.pyplot as plt
import pandas as pd
import re

In [2]:
path_to_data = 'nuggets5k.json' # use smaller file for development              
tweet_data = []
tweet_file = open(path_to_data, 'r')

In [3]:
for line in tweet_file:
    try:
        tweet = json.loads(line)
        tweet_data.append(tweet)
    except:
        continue

print len(tweet_data)

5007


There are 5007 total tweets that were stored in the `nuggets5k.json` file by the `twitter_mine.py` program.

# Extract and organize data

Examine a tweet to see what kind of data are available 

In [17]:
tweet_data[11]

{u'contributors': None,
 u'coordinates': None,
 u'created_at': u'Wed Sep 28 18:18:31 +0000 2016',
 u'entities': {u'hashtags': [],
  u'symbols': [],
  u'urls': [{u'display_url': u'bit.ly/2dDAEap',
    u'expanded_url': u'http://bit.ly/2dDAEap',
    u'indices': [59, 82],
    u'url': u'https://t.co/FWJJjww3oT'}],
  u'user_mentions': []},
 u'favorite_count': 0,
 u'favorited': False,
 u'filter_level': u'low',
 u'geo': None,
 u'id': 781196404836278273,
 u'id_str': u'781196404836278273',
 u'in_reply_to_screen_name': None,
 u'in_reply_to_status_id': None,
 u'in_reply_to_status_id_str': None,
 u'in_reply_to_user_id': None,
 u'in_reply_to_user_id_str': None,
 u'is_quote_status': False,
 u'lang': u'und',
 u'place': None,
 u'possibly_sensitive': False,
 u'retweet_count': 0,
 u'retweeted': False,
 u'source': u'<a href="http://twitterfeed.com" rel="nofollow">twitterfeed</a>',
 u'text': u'Microsoft y Lenovo impiden instalar Linux en... - PoderPDA https://t.co/FWJJjww3oT',
 u'timestamp_ms': u'147508671

To begin, extract the following fields into a data file:
* text (the actual content of the tweet -- for which some NLP may be done)
* lang
* location
* hashtags (keywords, can serve as proxy for the content subject)
* retweeted (boolean -- is this content worth repeating?)
* followers_count (how many people is this content reaching?)
* friends_count (how many people is this content reaching?)
* name (how many unique users posting about these topics?)

In [50]:
def add_column(keys, col_name):
    data = []
    
    for tweet in tweet_data:
        if len(keys) == 1:
            if tweet.has_key(keys[0]):
                data.append(tweet[keys[0]])
            else:
                data.append(None)
        elif len(keys) == 2:
            if tweet.has_key(keys[0]) and tweet[keys[0]].has_key(keys[1]):
                data.append(tweet[keys[0]][keys[1]])
            else:
                data.append(None)
        else: 
            print 'Function not designed to handle key depth > 2'
            return
    df[col_name] = data

In [87]:
df = pd.DataFrame()
add_column(['text'], 'text')
add_column(['user', 'lang'], 'user_lang')
add_column(['user', 'location'], 'location')
add_column(['entities', 'hashtags'], 'hashtags')
add_column(['retweeted'], 'retweet')
add_column(['retweet_count'], 'n_retweet')
add_column(['user', 'followers_count'], 'n_followers')
add_column(['user', 'friends_count'], 'n_friends')
add_column(['user', 'name'], 'name')
add_column(['coordinates'], 'coords')

In [76]:
df.head()

text user_lang  \
0  RT @JessikaDommeUK: I am now available for #Di...        en   
1  yNewsbot: Bugtraq: #Cisco #Security #Advisory:...        en   
2  RT @mims: Amazon is going to compete directly ...        en   
3  #Cisco Battles Shadow Broker Exploits. Read mo...        en   
4  #Cisco Battles Shadow Broker Exploits. Read mo...        en   

                  location                                           hashtags  \
0                       UK  [{u'indices': [43, 54], u'text': u'DirectChat'...   
1  tryingto #^fi~{|__#}%%*  [{u'indices': [19, 25], u'text': u'Cisco'}, {u...   
2                     None                                                 []   
3            San Mateo, CA          [{u'indices': [0, 6], u'text': u'Cisco'}]   
4            San Mateo, CA          [{u'indices': [0, 6], u'text': u'Cisco'}]   

  retweet  n_retweet  n_followers  n_friends                  name coords  
0   False          0        12633        933  ⭐️AdultworksFinest⭐️   None  
1   False          0         2636       1084      ❌Siraj Solution❌   None  
2   False          0         2624       4916    Leroy Wesley Annon   None  
3   False          0           86          2       Networking News   None  
4   False          0           20          2    Communication News   None

Latitude/longitude coordinates are potentially of interest to some applications.  Unfortunately, they exist for < 1% of the data.

In [94]:
coords = [cd for cd in df['coords'] if cd is not None]
print len(coords)
print len(df['coords'])
print float(len(coords)) / len(df['coords'])
for c in coords:
    print c, '\n'
    

15
5007
0.00299580587178
{u'type': u'Point', u'coordinates': [-121.8863286, 37.3382082]} 

{u'type': u'Point', u'coordinates': [-84.3879824, 33.7489954]} 

{u'type': u'Point', u'coordinates': [7.5949121, 50.3680029]} 

{u'type': u'Point', u'coordinates': [-95.6090009, 29.7801889]} 

{u'type': u'Point', u'coordinates': [-82.21336234, 27.03644926]} 

{u'type': u'Point', u'coordinates': [-98.4936282, 29.4241219]} 

{u'type': u'Point', u'coordinates': [-76.7238, 39.192978]} 

{u'type': u'Point', u'coordinates': [-78.898619, 35.9940329]} 

{u'type': u'Point', u'coordinates': [-91.4894045, 41.6838135]} 

{u'type': u'Point', u'coordinates': [-97.7430608, 30.267153]} 

{u'type': u'Point', u'coordinates': [-1.5359478, 53.7988358]} 

{u'type': u'Point', u'coordinates': [-1.5359478, 53.7988358]} 

{u'type': u'Point', u'coordinates': [-87.6297982, 41.8781136]} 

{u'type': u'Point', u'coordinates': [-77.1772604, 38.9338676]} 

{u'type': u'Point', u'coordinates': [-122.2284532, 47.3073228]} 



# (Some) Text Processing

In [96]:
def word_in_text(word, text):
    '''                                                                         
    Search for word in text (case insensitive), return True if found, else      
    return False                                                                
    '''
    match =  re.search(word.lower(), text.lower())
    if match:
        return True
    return False

In [106]:
# Test                                                                          
ex_text = 'This is a text string.'  
print 'text:', text
print '"this" in text:', word_in_text('this', text)                            
print '"TEXT" in text:', word_in_text('TEXT', text)                            
print '"word" in text:', word_in_text('word', text)                            

text: This is a text string.
"this" in text: True
"TEXT" in text: True
"word" in text: False


In [107]:
keywords = ['Cisco', 'CCNA', 'AWS', 'Sharepoint', 'linux', 'powershell',
            'itil', 'sql', 'azure', 'ceh']

In [127]:
for word in keywords:
    df[word] = df['text'].apply(lambda tweet: word_in_text(word, tweet)
                                if tweet else None)
    
df.head()

text user_lang  \
0  RT @JessikaDommeUK: I am now available for #Di...        en   
1  yNewsbot: Bugtraq: #Cisco #Security #Advisory:...        en   
2  RT @mims: Amazon is going to compete directly ...        en   
3  #Cisco Battles Shadow Broker Exploits. Read mo...        en   
4  #Cisco Battles Shadow Broker Exploits. Read mo...        en   

                  location                                           hashtags  \
0                       UK  [{u'indices': [43, 54], u'text': u'DirectChat'...   
1  tryingto #^fi~{|__#}%%*  [{u'indices': [19, 25], u'text': u'Cisco'}, {u...   
2                     None                                                 []   
3            San Mateo, CA          [{u'indices': [0, 6], u'text': u'Cisco'}]   
4            San Mateo, CA          [{u'indices': [0, 6], u'text': u'Cisco'}]   

  retweet  n_retweet  n_followers  n_friends                  name coords  \
0   False          0        12633        933  ⭐️AdultworksFinest⭐️   None   
1   False          0         2636       1084      ❌Siraj Solution❌   None   
2   False          0         2624       4916    Leroy Wesley Annon   None   
3   False          0           86          2       Networking News   None   
4   False          0           20          2    Communication News   None   

   Cisco   CCNA    AWS Sharepoint  linux powershell   itil    sql  azure  \
0  False  False  False      False  False      False  False  False  False   
1   True  False  False      False  False      False  False  False  False   
2  False  False   True      False  False      False  False  False  False   
3   True  False  False      False  False      False  False  False  False   
4   True  False  False      False  False      False  False  False  False   

     ceh  
0  False  
1  False  
2  False  
3  False  
4  False

Write DataFrame to .csv for potential further processing
Note: encoding defaults to ascii, but will throw error if non-ascii text  found (e.g., international text), hence the utf-8 encoding argument.

In [128]:
df.to_csv('twitter.csv', encoding = 'utf-8')

In [129]:
# Alternately pickle as a Python object
df.to_pickle('twitter_pickle')

Analyses are conducted in a separate notebook.